File: mac/Sources/VibeBoardKit/Protocol/ControlCommand.swift
import Foundation

public enum VoiceKey: String, CaseIterable, Sendable {
    case none
    case k1
    case k2
    case k3
    case k4
}

public enum InteractionMode: String, CaseIterable, Sendable {
    case holdToTalk = "hold_to_talk"
    case clickToTalk = "click_to_talk"
}

public enum DeviceUIState: String, CaseIterable, Sendable {
    case ready
    case thinking
    case listening
    case processing
    case error
}

public enum ControlCommand: Equatable, Sendable {
    case announceUSBTransport
    case getDeviceInfo
    case uiState(DeviceUIState, text: String)
    case ping
    case interactionMode(InteractionMode)
    case voiceKey(VoiceKey)
}

public enum FrameEncoder {
    public static func encode(_ command: ControlCommand) throws -> Data {
        let body: Data

        switch command {
        case .announceUSBTransport:
            body = Data(#"{"event":"transport","kind":"usb"}"#.utf8)
        case .getDeviceInfo:
            body = Data(#"{"event":"get_device_info"}"#.utf8)
        case .uiState(let state, let text):
            body = Data(
                (#"{"event":"ui_state","state":""#
                    + state.rawValue
                    + #"","text":""#
                    + jsonEscapedContent(text)
                    + #""}"#).utf8
            )
        case .ping:
            body = Data(#"{"event":"ping"}"#.utf8)
        case .interactionMode(let mode):
            body = Data(
                (#"{"event":"interaction_mode","mode":""#
                    + mode.rawValue
                    + #""}"#).utf8
            )
        case .voiceKey(let key):
            body = Data(
                (#"{"event":"voice_key","key":""#
                    + key.rawValue
                    + #""}"#).utf8
            )
        }

        return try encodeOrdinary(type: .state, body: body)
    }

    public static func encodeOrdinary(type: FrameType, body: Data) throws -> Data {
        guard type == .state else {
            throw ProtocolError.unsupportedFrameType(type.rawValue)
        }
        guard body.count <= Int(UInt16.max) else {
            throw ProtocolError.bodyLengthOutOfRange(body.count)
        }

        let totalLength = 4 + body.count
        guard totalLength <= FrameStreamParser.maximumFrameLength else {
            throw ProtocolError.frameTooLarge(
                actual: totalLength,
                maximum: FrameStreamParser.maximumFrameLength
            )
        }

        let length = UInt16(body.count)
        var frame = Data([
            FrameStreamParser.protocolVersion,
            type.rawValue,
            UInt8(truncatingIfNeeded: length),
            UInt8(truncatingIfNeeded: length >> 8)
        ])
        frame.append(body)
        return frame
    }

    private static func jsonEscapedContent(_ value: String) -> String {
        var result = ""
        result.reserveCapacity(value.utf8.count)

        for scalar in value.unicodeScalars {
            switch scalar.value {
            case 0x22:
                result += #"\""#
            case 0x5c:
                result += #"\\"#
            case 0x08:
                result += #"\b"#
            case 0x0c:
                result += #"\f"#
            case 0x0a:
                result += #"\n"#
            case 0x0d:
                result += #"\r"#
            case 0x09:
                result += #"\t"#
            case 0x00...0x1f:
                result += String(format: "\\u%04x", scalar.value)
            default:
                result.unicodeScalars.append(scalar)
            }
        }

        return result
    }
}

File: win/src/VibeBoardKit/Protocol/ControlCommand.cs
using System;
using System.Text;
using System.Text.Json;

namespace VibeBoardKit.Protocol;

/// <summary>
/// Voice key selection for the device microphone capture.
/// </summary>
public enum VoiceKey
{
    None,
    K1,
    K2,
    K3,
    K4
}

/// <summary>
/// Interaction mode for voice input on the device.
/// </summary>
public enum InteractionMode
{
    HoldToTalk,
    ClickToTalk
}

/// <summary>
/// UI state the host pushes to the device.
/// </summary>
public enum DeviceUIState
{
    Ready,
    Thinking,
    Listening,
    Processing,
    Error
}

/// <summary>
/// Commands the host can send to the device over the ordinary JSON state channel.
/// </summary>
public abstract class ControlCommand
{
    private ControlCommand() { }

    public sealed class AnnounceUSBTransport : ControlCommand { }
    public sealed class GetDeviceInfo : ControlCommand { }
    public sealed class UIState : ControlCommand
    {
        public DeviceUIState State { get; }
        public string Text { get; }
        public UIState(DeviceUIState state, string text) { State = state; Text = text; }
    }
    public sealed class Ping : ControlCommand { }
    public sealed class InteractionModeCmd : ControlCommand
    {
        public InteractionMode Mode { get; }
        public InteractionModeCmd(InteractionMode mode) { Mode = mode; }
    }
    public sealed class VoiceKeyCmd : ControlCommand
    {
        public VoiceKey Key { get; }
        public VoiceKeyCmd(VoiceKey key) { Key = key; }
    }
}

/// <summary>
/// Encodes <see cref="ControlCommand"/> instances into binary frames for USB transmission.
/// </summary>
public static class FrameEncoder
{
    public static byte[] Encode(ControlCommand command)
    {
        string body = command switch
        {
            ControlCommand.AnnounceUSBTransport => "{\"event\":\"transport\",\"kind\":\"usb\"}",
            ControlCommand.GetDeviceInfo => "{\"event\":\"get_device_info\"}",
            ControlCommand.UIState ui => $"{{\"event\":\"ui_state\",\"state\":\"{UIStateString(ui.State)}\",\"text\":\"{JsonEscape(ui.Text)}\"}}",
            ControlCommand.Ping => "{\"event\":\"ping\"}",
            ControlCommand.InteractionModeCmd im => $"{{\"event\":\"interaction_mode\",\"mode\":\"{InteractionModeString(im.Mode)}\"}}",
            ControlCommand.VoiceKeyCmd vk => $"{{\"event\":\"voice_key\",\"key\":\"{VoiceKeyString(vk.Key)}\"}}",
            _ => throw new ProtocolException($"Unknown command type: {command.GetType().Name}")
        };

        return EncodeOrdinary(FrameType.State, body);
    }

    public static byte[] EncodeOrdinary(FrameType type, string body)
    {
        if (type != FrameType.State)
            throw new ProtocolException($"Unsupported frame type for ordinary encoding: {type}");

        byte[] bodyBytes = Encoding.UTF8.GetBytes(body);
        if (bodyBytes.Length > ushort.MaxValue)
            throw new ProtocolException($"Body length out of range: {bodyBytes.Length}");

        int totalLength = 4 + bodyBytes.Length;
        if (totalLength > FrameStreamParser.MaximumFrameLength)
            throw new ProtocolException($"Frame too large: actual={totalLength}, max={FrameStreamParser.MaximumFrameLength}");

        var frame = new byte[totalLength];
        frame[0] = FrameStreamParser.ProtocolVersion;
        frame[1] = (byte)type;
        ushort length = (ushort)bodyBytes.Length;
        frame[2] = (byte)(length & 0xFF);
        frame[3] = (byte)((length >> 8) & 0xFF);
        Array.Copy(bodyBytes, 0, frame, 4, bodyBytes.Length);
        return frame;
    }

    public static string JsonEscape(string value)
    {
        var sb = new StringBuilder(value.Length);
        foreach (char c in value)
        {
            switch (c)
            {
                case '"': sb.Append("\\\""); break;
                case '\\': sb.Append("\\\\"); break;
                case '\b': sb.Append("\\b"); break;
                case '\f': sb.Append("\\f"); break;
                case '\n': sb.Append("\\n"); break;
                case '\r': sb.Append("\\r"); break;
                case '\t': sb.Append("\\t"); break;
                default:
                    if (c < 0x20)
                        sb.Append($"\\u{(int)c:X4}");
                    else
                        sb.Append(c);
                    break;
            }
        }
        return sb.ToString();
    }

    public static string UIStateString(DeviceUIState state) => state switch
    {
        DeviceUIState.Ready => "ready",
        DeviceUIState.Thinking => "thinking",
        DeviceUIState.Listening => "listening",
        DeviceUIState.Processing => "processing",
        DeviceUIState.Error => "error",
        _ => throw new ProtocolException($"Unknown UI state: {state}")
    };

    public static string InteractionModeString(InteractionMode mode) => mode switch
    {
        InteractionMode.HoldToTalk => "hold_to_talk",
        InteractionMode.ClickToTalk => "click_to_talk",
        _ => throw new ProtocolException($"Unknown interaction mode: {mode}")
    };

    public static string VoiceKeyString(VoiceKey key) => key switch
    {
        VoiceKey.None => "none",
        VoiceKey.K1 => "k1",
        VoiceKey.K2 => "k2",
        VoiceKey.K3 => "k3",
        VoiceKey.K4 => "k4",
        _ => throw new ProtocolException($"Unknown voice key: {key}")
    };
}
